### PR TITLE
bugfix/RLP-836/cannot-receive-payments

### PR DESCRIPTION
**Pull request by mauricioirace**
Change log:
- Broadcast some channel state changes to multiple channels,to prevent issuees caused by channel duplication

### CODE REVIEW

File: raiden/transfer/token_network.py [hunk 1]
```diff
@@ -41,18 +44,27 @@ def subdispatch_to_channel_by_id_and_address(
...
     ids_to_channels = token_network_state.channelidentifiers_to_channels
 
-    channel_state = None
+    channel_states = []
     if node_address is not None:
         if node_address in ids_to_channels:
             channel_state = ids_to_channels[node_address].get(state_change.channel_identifier)
+            channel_states.append(channel_state)
+            if channel_state.both_participants_are_light_clients\
+                and type(state_change) in DuplicableStateChanges:
```
Comment (mauricioirace) [inline raiden/transfer/token_network.py]: This could be encapsulaed

File: raiden/transfer/token_network.py [hunk 2]
```diff
@@ -41,18 +44,27 @@ def subdispatch_to_channel_by_id_and_address(
...
+                channel_states.append(partner_channel_state)
         else:
             lc_address = views.get_lc_address_by_channel_id_and_partner(token_network_state, node_address,
                                                                     state_change.canonical_identifier)
             node_address = lc_address
             if lc_address in token_network_state.channelidentifiers_to_channels:
                 channel_state = token_network_state.channelidentifiers_to_channels[lc_address].get(
                     state_change.canonical_identifier.channel_identifier)
-        if channel_state:
+                channel_states.append(channel_state)
```
Comment (mauricioirace) [inline raiden/transfer/token_network.py]: This could be also extracted into a method

File: raiden/transfer/token_network.py [hunk 3]
```diff
@@ -41,18 +44,27 @@ def subdispatch_to_channel_by_id_and_address(
...
+            if channel_state.both_participants_are_light_clients\
+                and type(state_change) in DuplicableStateChanges:
+                partner_channel_state = token_network_state.channelidentifiers_to_channels[channel_state.partner_state.address].get(
+                    channel_state.identifier
+                )
+                channel_states.append(partner_channel_state)
         else:
             lc_address = views.get_lc_address_by_channel_id_and_partner(token_network_state, node_address,
                                                                     state_change.canonical_identifier)
             node_address = lc_address
```
Comment (mauricioirace) [inline raiden/transfer/token_network.py]: I think we can delete this line now

File: raiden/transfer/token_network.py [hunk 4]
```diff
@@ -29,7 +29,10 @@
     ContractReceiveUpdateTransfer,
 ]
 
-
+# State changes that should be duplicated in case there are two light clients
+DuplicableStateChanges = [
```
Comment (mauricioirace) [inline raiden/transfer/token_network.py]: This is sort of a constant, not a class

Comment (mortelli) [reply raiden/transfer/token_network.py]: do we expect this list to grow?

File: raiden/transfer/token_network.py (hunk 3 above)
Comment (mortelli) [reply raiden/transfer/token_network.py]: with the code as it is now, yeah, it's unused

File: raiden/transfer/token_network.py [hunk 5]
```diff
@@ -41,18 +44,27 @@ def subdispatch_to_channel_by_id_and_address(
...
+    channel_states = []
     if node_address is not None:
         if node_address in ids_to_channels:
             channel_state = ids_to_channels[node_address].get(state_change.channel_identifier)
+            channel_states.append(channel_state)
+            if channel_state.both_participants_are_light_clients\
+                and type(state_change) in DuplicableStateChanges:
+                partner_channel_state = token_network_state.channelidentifiers_to_channels[channel_state.partner_state.address].get(
+                    channel_state.identifier
+                )
```
Comment (mortelli) [inline raiden/transfer/token_network.py]: this indentation is too confusing and the line too long, may i suggest:

```suggestion
                partner_channel = token_network_state.channelidentifiers_to_channels[channel_state.partner_state.address]
                partner_channel_state = partner_channel.get(channel_state.identifier)
```

**Review by mortelli — COMMENTED** [review]
i have read issue RLP-836 and taken a look at this PR, but still don't understand why this is the solution.

in particular, this statement is unclear to me:
> State changes that should be duplicated in case there are two light clients

as well as the `if` statement which uses this value.

could you please explain why this is the fix?

an high-level overview of the solution would be a huge help :) 

thanks!

File: raiden/transfer/token_network.py (hunk 4 above)
Comment (marcosmartinez7) [inline raiden/transfer/token_network.py]: I think we can rename this to represent that "we are dispatching to multiple channels because the channel is related with two light clients of the same hub"

File: raiden/transfer/token_network.py [hunk 6]
```diff
@@ -29,7 +29,10 @@
     ContractReceiveUpdateTransfer,
 ]
 
-
+# State changes that should be duplicated in case there are two light clients
+DuplicableStateChanges = [
+    ContractReceiveChannelNewBalance
+]
 def subdispatch_to_channel_by_id_and_address(
```
Comment (marcosmartinez7) [inline raiden/transfer/token_network.py]: i think we can invoke another funcion `subdispatch_to_channel_list `or something, to subdispatch a state change to multiple channels

File: raiden/transfer/token_network.py [hunk 7]
```diff
@@ -30,50 +30,78 @@
...
     token_network_state: TokenNetworkState,
     state_change: StateChangeWithChannelID,
     block_number: BlockNumber,
     block_hash: BlockHash,
-    node_address: AddressHex = None
+    participant_address: AddressHex = None
 ) -> TransitionResult:
-    events = list()
+    if participant_address is None:
+        return TransitionResult(token_network_state, [])
```
Comment (mortelli) [inline raiden/transfer/token_network.py]:
```suggestion
    if not participant_address:
        return TransitionResult(token_network_state, [])
```

File: raiden/transfer/token_network.py [hunk 8]
```diff
@@ -30,50 +30,78 @@
...
+    block_hash: BlockHash,
+    block_number: BlockNumber,
+    channel_states: List[NettingChannelState],
+    state_change: StateChangeWithChannelID,
+    token_network_state: TokenNetworkState
+) -> TransitionResult:
+    events = []
     ids_to_channels = token_network_state.channelidentifiers_to_channels
+    for channel_state in channel_states:
+        result = channel.state_transition(
```
Comment (mortelli) [inline raiden/transfer/token_network.py]: i would name this something more explicit like `channel_result`

File: raiden/transfer/token_network.py [hunk 9]
```diff
@@ -30,50 +30,78 @@
...
-                partner_to_channelids.remove(channel_identifier)
-            else:
-                ids_to_channels[node_address][channel_identifier] = result.new_state
+        events.extend(result.events)
+    return TransitionResult(token_network_state, events)
 
-            events.extend(result.events)
 
-    return TransitionResult(token_network_state, events)
+#  TODO: We should find another approach for ths function. It's "a bit" confusing
```
Comment (mortelli) [inline raiden/transfer/token_network.py]: either we deal with this or we remove it.

if necessary, please create a ticket. 

there is not much added value for this comment, i think

File: raiden/transfer/token_network.py [hunk 10]
```diff
@@ -30,50 +30,78 @@
...
+    ids_to_channels = token_network_state.channelidentifiers_to_channels
+
+    if participant_address in ids_to_channels:
+        channel_state = ids_to_channels[participant_address].get(state_change.channel_identifier)
+        channel_states.append(channel_state)
+        if channel_state.both_participants_are_light_clients:
+            partner_channel = ids_to_channels[channel_state.partner_state.address]
+            partner_channel_state = partner_channel.get(channel_state.identifier)
+            channel_states.append(partner_channel_state)
+    else:
```
Comment (mortelli) [inline raiden/transfer/token_network.py]: the code structure looks good, but i don't understand what this `if` is evaluating.

what does it mean that `participant_address in ids_to_channels == True`?

can we maybe express this boolean condition in a clearer, more business logic-oriented way?

at first i thought this was about light clients, but it seems both branches talk about LCs, so i'm not so sure now

**Review by mortelli — COMMENTED** [review]
it's much better now!

thanks for iterating this :smiley:

File: raiden/transfer/token_network.py (hunk 10 above)
Comment (mauricioirace) [reply raiden/transfer/token_network.py]: This was part of the discussion we had yesterday.

It's seems like both branches can be lc or full nodes, even the else branch can be a full node. I found the naming wrong if that's the case

I understand that:
If the participant is our (we are the full node owner of the address, or the corresponding lc is handled by us) then it goes to the first branch.

Else, it check if we have a channel with that address as a partner.

I think I know a few ways to improve this, but we don't even have an agreement on what does the 'else' does.

Comment (mortelli) [reply raiden/transfer/token_network.py]: > This was part of the discussion we had yesterday.
> 
> It's seems like both branches can be lc or full nodes, even the else branch can be a full node. I found the naming wrong if that's the case
> 
> I understand that:
> If the participant is our (we are the full node owner of the address, or the corresponding lc is handled by us) then it goes to the first branch.
> 
> Else, it check if we have a channel with that address as a partner.
> 
> I think I know a few ways to improve this, but we don't even have an agreement on what does the 'else' does.

if that's the case, could we name the condition something which reflects this?

e.g. `if participant_is_ours` or something like this?

Comment (mauricioirace) [reply raiden/transfer/token_network.py]: Agree.

Comment (mauricioirace) [reply raiden/transfer/token_network.py]: the other option I thought was renaming `ids_to_channels` to something like `our_channels`, so `participant_address in our_channels` would be a more explicit way of saying 'Is this participant address a side of some channel we own?'. 

this  would apply both to this condition and to the one in line 99

File: raiden/transfer/token_network.py [hunk 11]
```diff
@@ -30,64 +30,90 @@
 ]
 
 
-def subdispatch_to_channel_by_id_and_address(
+def subdispatch_to_channels_by_participant_address(
```
Comment (marcosmartinez7) [inline raiden/transfer/token_network.py:33]: I like this because: 

- we were not using channel id
- participant seems better than address

but.. im a little bit concern about the usage, we should test basically all the onchain operations manually and run the test suite

File: raiden/transfer/token_network.py [hunk 12]
```diff
@@ -30,64 +30,90 @@
...
 
-    return TransitionResult(token_network_state, events)
+def get_channels_to_dispatch_statechange(
+    participant_address: AddressHex,
+    state_change: StateChangeWithChannelID,
+    token_network_state: TokenNetworkState
+) -> List[NettingChannelState]:
+    channel_states = []
+    ids_to_channels = token_network_state.channelidentifiers_to_channels
+    participant_is_ours = participant_address in ids_to_channels
```
Comment (marcosmartinez7) [inline raiden/transfer/token_network.py:98]: maybe a comment that participant_is_ours represents ¨is a handled lc or is the hub itself¨ would be nice here

File: raiden/transfer/token_network.py [hunk 13]
```diff
@@ -30,64 +30,90 @@
...
-                partner_to_channelids.remove(channel_identifier)
-            else:
-                ids_to_channels[node_address][channel_identifier] = result.new_state
+        events.extend(channel_result.events)
+    return TransitionResult(token_network_state, events)
 
-            events.extend(result.events)
 
-    return TransitionResult(token_network_state, events)
+def get_channels_to_dispatch_statechange(
```
Comment (marcosmartinez7) [inline raiden/transfer/token_network.py:80]: I think this function does what we need, but both the name andthe StateChangeWithChannelID doesnt. 

`StateChangeWithChannelID ` represents: "a set of state changes that must be subdispatched to two channels because the channels are from two onboarded light clients"

`get_channels_to_dispatch_statechange` logic is strongly related with two light clients of the hub, it basically return 1 or 2 channel states depending on `both_participants_are_light_clients` value. Basically, given an address and a channel id, we check if that channel is also into the state of another LC

I suggest to change StateChangeWithChannelID for MultipleChannelDispatchStateChange or something. 

About the function, the name is ok but we need docstring

**Review by marcosmartinez7 — CHANGES_REQUESTED** [review]

File: raiden/transfer/token_network.py (hunk 13 above)
Comment (JONAF2103) [reply raiden/transfer/token_network.py:80]: I Agree with @marcosmartinez7 here, i still don't like all this logic just to add 1 or 2 participant channel states to the list but i don't see any other way to do it, anyway adding a docstring and changing the state change name should help us to understand if we come back here later.

File: raiden/transfer/token_network.py (hunk 12 above)
Comment (mauricioirace) [reply raiden/transfer/token_network.py:98]: I'll comment 'is a handled lc or is the node itself'  to be more accurate

File: raiden/transfer/token_network.py (hunk 11 above)
Comment (mauricioirace) [reply raiden/transfer/token_network.py:33]: Agree, I'll run the tests at the end of the day. 
The method is only used internally in this file and seems like it hadn't break anythink that wasn't already broken (channel close and update balance proof) but would be wrong not to run the test suite

File: raiden/transfer/token_network.py (hunk 13 above)
Comment (mauricioirace) [reply raiden/transfer/token_network.py:80]: StateChangeWithChannelID was introduced a long time ago. It represents state changes that hold a channel identifier. I thik MultipleChannelDispatchStateChange would lost the meaning of the type

**Review by marcosmartinez7 — APPROVED** [review]
lgtm now, thanks!

**Review by marcosmartinez7 — APPROVED** [review]

File: raiden/transfer/token_network.py [hunk 14]
```diff
@@ -30,50 +30,88 @@
...
+def subdispatch_to_channels_by_participant_address(
     token_network_state: TokenNetworkState,
     state_change: StateChangeWithChannelID,
     block_number: BlockNumber,
     block_hash: BlockHash,
-    node_address: AddressHex = None
+    participant_address: AddressHex = None
 ) -> TransitionResult:
-    events = list()
+    if not participant_address:
```
Comment (mortelli) [inline raiden/transfer/token_network.py:40]: it seems a bit strange to me that a caller would omit the `participant_address` parameter if this method is just going to return an empty `TransitionResult`. but i guess this is out of scope for this PR

File: raiden/transfer/token_network.py [hunk 15]
```diff
@@ -30,50 +30,88 @@
...
+    """
+    Retrieve which channels the state_change should be dispatched to.
+    In most situations, the state_change should be dispatched to this node's side of the channel.
+    If this node runs in hub mode, and both sides of the channel are light client handled by this hub,
+    then it's needed to dispatch the state_change to both channel_states (one per light client)
+    :param participant_address: address of the participan that fired the state_change
+    :param state_change: state_change to dispatch
+    :param token_network_state: token network where to look for channels
+    :return: Channel where  the state_change should be dispatched
+    """
```
Comment (mortelli) [inline raiden/transfer/token_network.py]: great :clap:

File: raiden/transfer/token_network.py [hunk 16]
```diff
@@ -30,50 +30,88 @@
...
+) -> List[NettingChannelState]:
+    """
+    Retrieve which channels the state_change should be dispatched to.
+    In most situations, the state_change should be dispatched to this node's side of the channel.
+    If this node runs in hub mode, and both sides of the channel are light client handled by this hub,
+    then it's needed to dispatch the state_change to both channel_states (one per light client)
+    :param participant_address: address of the participan that fired the state_change
+    :param state_change: state_change to dispatch
+    :param token_network_state: token network where to look for channels
+    :return: Channel where  the state_change should be dispatched
```
Comment (mortelli) [inline raiden/transfer/token_network.py]: - missed a `t` in `participan`
- extra space in `where  the`

File: raiden/transfer/token_network.py [hunk 17]
```diff
@@ -30,50 +30,88 @@
...
+    """
+    channel_states = []
+    ids_to_channels = token_network_state.channelidentifiers_to_channels
+    # is a handled lc or is the node itself'
```
Comment (mortelli) [inline raiden/transfer/token_network.py]: is the `'` supposed to be a `?`?

**Review by mortelli — APPROVED** [review]
you got the green light from me as long as the UTs and MTs turn out OK :+1:

**Review by JONAF2103 — APPROVED** [review]
LGTM, you should update the docs with the typo errors.
